pkg/engine2/operational_eval/vertex_resource_rule_test.go

```diff
@@ -0,0 +1,220 @@
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			got := tt.v.Key()
+			assert.Equal(tt.want, got)
+		})
+	}
+}
+
+func Test_resourceRuleVertex_Dependencies(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	dcap := NewMockdependencyCapturer(ctrl)
+
+	tests := []struct {
+		name     string
+		resource construct.ResourceId
+		rule     knowledgebase.AdditionalRule
+		mocks    func()
+		wantErr  bool
+	}{
+		{
+			name:     "resource rule vertex",
+			resource: construct.ResourceId{Name: "test"},
+			rule: knowledgebase.AdditionalRule{
+				If: "test",
+			},
+			mocks: func() {
+				dcap.EXPECT().ExecuteOpRule(knowledgebase.DynamicValueData{
+					Resource: construct.ResourceId{Name: "test"},
+				}, knowledgebase.OperationalRule{
+					If: "test",
+				}).Return(nil)
+			},
+		},
+		{
+			name:     "resource rule vertex dependencies throws error",
+			resource: construct.ResourceId{Name: "test"},
+			rule: knowledgebase.AdditionalRule{
+				If: "test",
+			},
+			mocks: func() {
+				dcap.EXPECT().ExecuteOpRule(knowledgebase.DynamicValueData{
+					Resource: construct.ResourceId{Name: "test"},
+				}, knowledgebase.OperationalRule{
+					If: "test",
+				}).Return(fmt.Errorf("err"))
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			v := &resourceRuleVertex{
+				Resource: tt.resource,
+				Rule:     tt.rule,
+			}
+			tt.mocks()
+			eval := &Evaluator{}
+			err := v.Dependencies(eval, dcap)
+			if tt.wantErr {
+				assert.Error(err)
+				return
+			}
+			assert.NoError(err)
+			dcap.ctrl.Finish()
+		})
+	}
+}
+
+func Test_resourceRuleVertex_UpdateFrom(t *testing.T) {
+
+	tests := []struct {
+		name    string
+		initial *resourceRuleVertex
+		other   Vertex
+		want    resourceRuleVertex
+		wantErr bool
+	}{
+		{
+			name: "empty resource rule vertex",
+			initial: &resourceRuleVertex{
+				Resource: construct.ResourceId{Name: "test"},
+			},
+			other: &resourceRuleVertex{
+				Resource: construct.ResourceId{Name: "test"},
+				Rule: knowledgebase.AdditionalRule{
+					If: "test",
+				},
+			},
+			want: resourceRuleVertex{
+				Resource: construct.ResourceId{Name: "test"},
+				Rule: knowledgebase.AdditionalRule{
+					If: "test",
+				},
+			},
+		},
+		{
+			name:    "panics if different ref on resource rule vertex",
+			initial: &resourceRuleVertex{},
+			other: &resourceRuleVertex{
+				Resource: construct.ResourceId{Name: "test"},
+				Rule: knowledgebase.AdditionalRule{
+					If: "test",
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name:    "panics if not resource rule vertex",
+			initial: &resourceRuleVertex{},
+			other:   &propertyVertex{},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			defer func() {
+				if r := recover(); r != nil {
```

Comment:
Testing panics is just testing for coverage and doesn't actually give us any more confidence in the engine since a panic should never happen except by programming error. We don't test for index out of bounds or nil pointers either.